main/src/ca/uwaterloo/flix/language/phase/Resolver.scala

```diff
@@ -1746,7 +1746,8 @@ private def visitDef(defn: NamedAst.Declaration.Def, loc: SourceLocation)(implic
     */
 private def visitApplyDef(defn: NamedAst.Declaration.Def, exps: List[NamedAst.Expr], env: ListMap[String, Resolver.Resolution], innerLoc: SourceLocation, outerLoc: SourceLocation)(implicit scope: Scope, ns0: Name.NName, taenv: Map[Symbol.TypeAliasSym, ResolvedAst.Declaration.TypeAlias], root: NamedAst.Root, flix: Flix): Validation[ResolvedAst.Expr, ResolutionError] = {
     mapN(traverse(exps)(resolveExp(_, env))) {
-      es => visitApplyToplevelFull(ResolvedAst.Expr.Def(defn.sym, innerLoc), defn.spec.fparams.length, es, outerLoc)
+      case es if defn.spec.fparams.length == es.length => ResolvedAst.Expr.ApplyDef(Ast.DefSymUse(defn.sym, innerLoc), es, outerLoc)
+      case es => visitApplyToplevelFull(ResolvedAst.Expr.Def(defn.sym, innerLoc), defn.spec.fparams.length, es, outerLoc)
```

Comment:
This call will make an Apply(Def(..), ..) which shouldve been and ApplyDef. Essentially we should copy visitApplyToplevelFull twice, one where `base: DefSymUse` and one where `base: ResolvedAst.Expr.Sig`. these could stay as one function if we give a full application constructor function as `base: List[Exp] => Exp`.

Reply:
I see but right now we support both (in this pr)